homeassistant/helpers/template.py

```diff
@@ -487,6 +488,22 @@ def forgiving_float(value):
         return value
 
 
+def hsv_to_rgb(value):
```

Comment:
Sorry I didn't see those. There is no documentation about the bounds of each values. It seems from the tests H is from 0-65536 and others 0-255. Is that correct? Any particular reason? What if one wants 255 bounds for H?

Given that, it would make sense to have `convert_8_to_16` and friends as filters too as bitwise operations are not allowed by Jinja: https://github.com/pallets/jinja/issues/249

Reply:
Actually I'm not sure how to handle the bit issue as we work with arrays... Any approach I can think of would be bloated...

Reply:
If it's too bloated, maybe it shouldn't be part of templates. We can't solve every use case. Consider using things like Python scripts. You can just copy paste your functions in there and do whatever you want.

Reply:
AFAIK Python scripts cannot do imports.

Reply:
No need to import if you just copy paste the function into your script.

Reply:
I think a jinja filter is a more elegant solution. I've implemented the changes you requested, let me know if there's anything else.

Reply:
Elegant for your use case, taking up memory for everyone else that doesn't use HSV values. I'm not convinced yet that this would benefit from being part of the core experience. Especially since a workaround (copy pasting the method into Python scripts) also works.

Reply:
Components tplink, wink, yeelight, hue, mystorm and lifx use HSV internaly (to some degree) and even some use directly `colorsys`, bypassing `util.color`.

I think home assistant lights should be able to speak HSV, I'm surprised everyone comes up with its own hack and there is still no global solution for this.

In my opinion, a global solution would include:

1. `hsv_color` property (getter & setter) in light which may default to converting to rgb so all components are able to "speak hsv"
2. *This PR.* Templates functions to handle conversions (like for mqtt_template and others)

If you agree I can start working on the first point.

Reply:
@Diaoul Are both of those needed or would number one suffice?

HSV and RGB are equivalent (express the same thing) so balloob has a point that having both is a waste. However, HSV is much more user-friendly so I think it is unfortunate if we get stuck with RGB because it happened to show up first.

Reply:
[Here](https://github.com/FastLED/FastLED/issues/482#issuecomment-325029312) is a comment from a dev of [FastLED](https://github.com/FastLED/FastLED). He is rewriting a good part of the lib so we can work with RGB, HSV, 8 or 16bits at the high level without performance penalty.

Currently, Home Assistant is deeply bound to RGB 8 bit color representation and that's not such a bad thing, it's the most popular color representation after all.
However, I miss the ability to work with other representations which would allow some cool effects and make easier the color manipulation.
Home Assistant can afford a converting color from internal (RGB) representation to others as computational power isn't as limited as on a µC.
I'm also not a big fan of the current HSV converting functions as it is 16 bits for H and 8 bits for S and V...

What I suggest is add `hsv_color` (16 bits?) [here](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/light/__init__.py#L360-L368) and have it default to a property that converts back and forth (i.e. set and get) to `rgb_color`.
I'm not sure of the implications of this however.

Reply:
Would you mind answering my question?

Reply:
> Are both of those needed or would number one suffice?
Both what ? If functions then yes I need both if I want to increment hue
for example I need to get the current hue before setting it to an
incremented value.
If color representation, I guess at the low level one is enough but high
level multiple formats would suit most use cases.

Does that answer your question ? If not please elaborate.

Reply:
Sorry for the confusion. I don't mean to be an ass or ask rhetorical questions.

I was referring to your "global solution" with two points. My question was, is this PR (template filters) even needed if we get the `hsv_color` property?

Reply:
No it wouldn't be necessary to have both. `hsv_color` could be used on templates making the jinja filters useless.

Reply:
Okay, thanks. Since you said that you are prepared to work on the first point I assume that you think that the best solution actually is to get the `hsv_color` property?

In that case I suggest that you go ahead and make a new PR for that change. It might not get accepted but a PR is the best way to get feedback on a proposed change. Developers just cannot leave code alone :)

For what it is worth, I would also like hue support but I have no voting power for its inclusion.

(I guess an alternative is separate Hue and Saturation properties, to go along with the existing Brightness.)

Reply:
I like the idea of separate properties, that would make more sense right? Those could then be set via turn_on commands and integrations can just do whatever they want with it.

> For what it is worth, I would also like hue support but I have no voting power for its inclusion.

@amelchio, you are a member of the Home Assistant org and although we don't have "voting rights", your opinion definitely matters!

Reply:
Right, I just wanted to make it clear that even suggestions from members can be rejected (which is good).